File: snatch-rs/src/extractor.rs
use crate::models::{ExtractResponse, VideoFormat};
use crate::retry::{is_retryable_error, retry_with_backoff};
use crate::validation::{extract_platform, validate_url};
use std::process::Stdio;
use std::time::Duration;
use tokio::process::Command;
use tokio::time::timeout;

/// Extract video information using yt-dlp with retry logic
pub async fn extract_video_info(url: &str) -> Result<ExtractResponse, String> {
    // Validate URL first to prevent command injection
    validate_url(url)?;

    // Use retry logic for the extraction
    // Clone URL for the async closure
    let url = url.to_string();
    retry_with_backoff(|| extract_video_info_internal(&url), 3).await
}

/// Internal extraction function (can be retried)
async fn extract_video_info_internal(url: &str) -> Result<ExtractResponse, String> {
    // Run yt-dlp with JSON output and timeout
    let yt_dlp_timeout = Duration::from_secs(30);

    let output = timeout(yt_dlp_timeout, async {
        Command::new("yt-dlp")
            .args(["--dump-json", "--no-warnings", "--no-playlist", url])
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .output()
            .await
    })
    .await
    .map_err(|_| {
        "Request timeout (30s). The video may be too large or the server is busy.".to_string()
    })?
    .map_err(|e| format!("Failed to run yt-dlp: {}. Is yt-dlp installed?", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("yt-dlp error: {}", stderr.trim()));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    let json: serde_json::Value = serde_json::from_str(&stdout)
        .map_err(|e| format!("Failed to parse yt-dlp output: {}", e))?;

    // Extract platform using the validation module
    let platform = extract_platform(url);

    // Extract title
    let title = json["title"].as_str().unwrap_or("Untitled").to_string();

    // Extract thumbnail
    let thumbnail = json["thumbnail"].as_str().map(String::from);

    // Extract formats
    let formats = extract_formats(&json);

    Ok(ExtractResponse::new(platform, title, thumbnail, formats))
}

/// Extract video formats from yt-dlp JSON output
fn extract_formats(json: &serde_json::Value) -> Vec<VideoFormat> {
    let mut formats = Vec::new();

    // Try to get the best format directly
    if let Some(url) = json["url"].as_str() {
        formats.push(VideoFormat {
            quality: "best".to_string(),
            url: url.to_string(),
            ext: json["ext"].as_str().unwrap_or("mp4").to_string(),
            filesize: json["filesize"].as_u64(),
        });
        return formats;
    }

    // Parse formats array
    if let Some(format_list) = json["formats"].as_array() {
        // Filter video formats and sort by quality
        let mut video_formats: Vec<_> = format_list
            .iter()
            .filter(|f| {
                let vcodec = f["vcodec"].as_str().unwrap_or("none");
                vcodec != "none" && f["url"].as_str().is_some()
            })
            .collect();

        // Sort by height (quality) descending
        video_formats.sort_by(|a, b| {
            let h_a = a["height"].as_u64().unwrap_or(0);
            let h_b = b["height"].as_u64().unwrap_or(0);
            h_b.cmp(&h_a)
        });

        // Take top 3 qualities
        for f in video_formats.into_iter().take(3) {
            let height = f["height"].as_u64().unwrap_or(0);
            let quality = if height > 0 {
                format!("{}p", height)
            } else {
                f["format_note"].as_str().unwrap_or("unknown").to_string()
            };

            formats.push(VideoFormat {
                quality,
                url: f["url"].as_str().unwrap_or("").to_string(),
                ext: f["ext"].as_str().unwrap_or("mp4").to_string(),
                filesize: f["filesize"].as_u64(),
            });
        }
    }

    // Fallback: try requested_downloads
    if formats.is_empty() {
        if let Some(downloads) = json["requested_downloads"].as_array() {
            for d in downloads {
                if let Some(url) = d["url"].as_str() {
                    formats.push(VideoFormat {
                        quality: "best".to_string(),
                        url: url.to_string(),
                        ext: d["ext"].as_str().unwrap_or("mp4").to_string(),
                        filesize: d["filesize"].as_u64(),
                    });
                }
            }
        }
    }

    formats
}

File: apps/api/src/lib/extractor.ts
import type { ExtractResponse, VideoFormat } from "@snatch/shared";

/**
 * Extract video information using yt-dlp with retry logic
 */
export async function extractVideoInfo(url: string): Promise<ExtractResponse> {
	const proc = Bun.spawn(["yt-dlp", "--dump-json", "--no-warnings", "--no-playlist", url], {
		stdout: "pipe",
		stderr: "pipe",
	});

	const timeout = setTimeout(() => {
		proc.kill();
	}, 30_000);

	try {
		const exitCode = await proc.exited;
		clearTimeout(timeout);

		if (exitCode !== 0) {
			const stderr = await new Response(proc.stderr).text();
			throw new Error(`yt-dlp error: ${stderr.trim()}`);
		}

		const stdout = await new Response(proc.stdout).text();
		const json = JSON.parse(stdout);

		return parseYtDlpOutput(json, url);
	} finally {
		clearTimeout(timeout);
	}
}

/**
 * Download video using yt-dlp and stream to response
 */
export function downloadVideoStream(url: string): ReadableStream<Uint8Array> {
	const proc = Bun.spawn(
		["yt-dlp", "-o", "-", "--no-warnings", "--no-playlist", "-f", "best[ext=mp4]/best", url],
		{
			stdout: "pipe",
			stderr: "pipe",
		},
	);

	return new ReadableStream({
		async pull(controller) {
			const reader = proc.stdout.getReader();
			try {
				while (true) {
					const { done, value } = await reader.read();
					if (done) {
						controller.close();
						break;
					}
					controller.enqueue(value);
				}
			} catch (error) {
				controller.error(error);
			}
		},
		cancel() {
			proc.kill();
		},
	});
}

/**
 * Parse yt-dlp JSON output into ExtractResponse
 */
// biome-ignore lint/suspicious/noExplicitAny: yt-dlp JSON output is dynamic
function parseYtDlpOutput(json: any, url: string): ExtractResponse {
	const platform = detectPlatform(url);
	const title = json.title || "Untitled";
	const thumbnail = json.thumbnail || undefined;
	const formats = extractFormats(json);

	return {
		success: true,
		platform,
		title,
		thumbnail,
		formats,
	};
}

/**
 * Extract formats from yt-dlp JSON
 */
// biome-ignore lint/suspicious/noExplicitAny: yt-dlp JSON output is dynamic
function extractFormats(json: any): VideoFormat[] {
	const formats: VideoFormat[] = [];

	// Try direct URL first
	if (json.url) {
		formats.push({
			quality: "best",
			url: json.url,
			ext: json.ext || "mp4",
			filesize: json.filesize || undefined,
		});
		return formats;
	}

	// Parse formats array
	if (json.formats?.length) {
		const videoFormats = json.formats
			// biome-ignore lint/suspicious/noExplicitAny: yt-dlp JSON is dynamic
			.filter((f: any) => f.vcodec !== "none" && f.url)
			// biome-ignore lint/suspicious/noExplicitAny: yt-dlp JSON is dynamic
			.sort((a: any, b: any) => (b.height || 0) - (a.height || 0))
			.slice(0, 3);

		for (const f of videoFormats) {
			const height = f.height || 0;
			formats.push({
				quality: height > 0 ? `${height}p` : f.format_note || "unknown",
				url: f.url,
				ext: f.ext || "mp4",
				filesize: f.filesize || undefined,
			});
		}
	}

	// Fallback: requested_downloads
	if (formats.length === 0 && json.requested_downloads?.length) {
		for (const d of json.requested_downloads) {
			if (d.url) {
				formats.push({
					quality: "best",
					url: d.url,
					ext: d.ext || "mp4",
					filesize: d.filesize || undefined,
				});
			}
		}
	}

	return formats;
}

/**
 * Detect platform from URL
 */
function detectPlatform(url: string): string {
	const host = (() => {
		try {
			return new URL(url).hostname.toLowerCase();
		} catch {
			return "";
		}
	})();

	if (host.includes("instagram.com")) return "instagram";
	if (host.includes("tiktok.com")) return "tiktok";
	if (host.includes("x.com") || host.includes("twitter.com")) return "twitter";
	return "unknown";
}
